### PR TITLE
Fix #966 - Remove Legacy NGSI-v1 from codebase

### LINKED ISSUE #966: Question: How/when to remove NGSI-v1 from the library?
I know **NGSI-v1** is deprecated, but at some point I assume **NGSI-v1** support is to be dropped. Just to ensure things don't get lost here is a reminder for future development:

1) [x] Delete NGSI-v1 modules refactored out by #842 
2) [x] Delete NGSI-v1 unit tests - look for prefix _NGSI-v1_
3) [ ] Look for instances of `isCurrentNgsi()` across current  IoT Agent codebases - see #963 - remove NGSI-v1 path of execution
4) [x]  Fix documentation (moving examples to NGSI-v2, remove NGSI-v1 specific topics, etc.)
5) [x]  Do a final check (grep-wise) for “NGSIv1”, “NGSI-v1”, “NGSI v1”, "v1", etc. to remove any additional leftover that could remain after the accomplished above steps

### Questions

- Anything missing with this course of action?
- This would of course be a **breaking change** - would the package move to `3.0.0`?
- Should this update be aligned with full **NGSI-LD** support being completed?

### PR DESCRIPTION
**Pull request by jason-fox**
- Remove `config.isCurrentNGSI()` clauses from plugins
- Remove NGSI-v1 Templates
- Remove V1 options from service
- Delete NGSI-V1 classes
- Delete  NGSI-V1 tests
- Convert Keystone and Registry tests to NGSI-v2

### CODE REVIEW

**Comment by jason-fox** [comment]
Related #966

**Comment by jason-fox** [comment]
@AlvaroVega @fgalan - this is now ready for review. The remaining reduction of 1.5% code coverage **may** still be NGSI-v1 lava code somewhere or it may be that some additional test cases need to be added (particularly around plugins). Either way I wouldn't feel comfortable modifying the code further as part of this PR.

**Comment by fgalan** [comment]
We are going to start testing on this. To that end, the following (temporal) PRs has been created:

* https://github.com/telefonicaid/iotagent-json/pull/533
* https://github.com/telefonicaid/iotagent-ul/pull/467
* https://github.com/telefonicaid/iotagent-manager/pull/238

While unit tests have gone fine in the IOTA Manager PR, they are failing in IOTA-UL and IOTA-JSON. Looking at the results, it seems it is not the usual glitch in some AMQP tests, but a massive fail. At least 75 tests are failing in the IOTA-UL case and 80 in the IOTA-JSON case.

Maybe I'm doing something wrong... @jason-fox could you have a look, please?

**Comment by mapedraza** [comment]
> @AlvaroVega @fgalan - this is now ready for review. The remaining reduction of 1.5% code coverage **may** still be NGSI-v1 lava code somewhere or it may be that some additional test cases need to be added (particularly around plugins). Either way I wouldn't feel comfortable modifying the code further as part of this PR.

@jason-fox Just to clarify, the idea is to add more test on a separate PR to increase the code coverage that 1.5% or leave it like this?

**Comment by jason-fox** [comment]
@mapedraza - leave. All I've done is delete the NGSI-v1 tests and copied over the obvious _tests for non NGSI functionality_ to use  NGSI-v2 requests and responses. I am uncertain where the remaining drop of 1.5% lies - I'm guessing various plugins need to have some NGSI-v1 cleaning to remove dead code, but it could be a coveralls accounting error.

**Comment by jason-fox** [comment]
>  At least 75 tests are failing in the IOTA-UL case and 80 in the IOTA-JSON case.
> Maybe I'm doing something wrong

The failing tests are provisioned using NGSI-v1. For example

```
      4) should send one update context per measure group to the Context Broker
time=2021-03-05T10:23:21.547Z | lvl=FATAL | corr=n/a | trans=n/a | op=IoTAgentNGSI.Global | msg=An unexpected exception has been raised. Ignoring: AssertionError [ERR_ASSERTION]: Mocks not yet satisfied:
POST http://192.168.1.1:1026/v1/updateContext
```

Has a mock which assumes that the IoT Agent upserts a provisioned device using the **deprecated** `/v1/updateContext` endpoint. This is no longer occurs since the default configuration is NGSI-v2. Just delete the NGSI-v1 tests and re-run with just the tests in the `test/ngsiv2`  folder. (I had to do this with the code in the PR as well)

**Comment by jason-fox** [comment]
It should be the case that if the NGSI-v1 tests are deleted for IOTA-JSON and IOT-UL there should not be a significant reduction of coverage. If coverage is reduced then some of the NGSI-v1 tests will need to be converted to NGSI-v2.

**Comment by jason-fox** [comment]
Example run for IOTA-JSON with no NGSI-v1 tests : https://github.com/jason-fox/iotagent-json/runs/2041091462?check_suite_focus=true

**Comment by fgalan** [comment]
Take into account that https://github.com/telefonicaid/iotagent-node-lib/pull/1034 adds some files related with NGSIv1:

* test/unit/examples/contextRequests/updateContextExpressionPlugin6.json
* test/unit/examples/contextRequests/updateContextExpressionPlugin7.json
* test/unit/examples/contextRequests/updateContextExpressionPlugin8.json

These ones should be removed in this PR upon conflict resolution when updated with master.

**Comment by jason-fox** [comment]
@fgalan  - seeing no movement on this, I've raised two new PRs to try and reduce the scope.

- If you action #1042 then no tests will run on NGSI-v1, and this PR will reduce by 189 files - coverage has temporarily reduced  by 8% with the test removal since all the legacy NGSI-v1 code remains, but it should return the same as here once this PR is finally actioned and NGSI-v1 deleted.

- If you action #1043 then no references to NGSI-v1 will remain in the docs, and this PR will reduce by 7 files - coverage should remain the same as here.

The following are currently  on your side:

- I assume the blocker on actually updating the code in the library  is here: https://github.com/telefonicaid/iotagent-json/issues/563 so you can successfully run the test suite in the JSON IoT Agent without NGSI-v1. Someone needs to decide whether the code or the test is in error and I guess that is on TEF.

- LightweightM2M test suite is currently broken on **master** - I assume this is due to some IoT Agent node lib change.
https://github.com/telefonicaid/lightweightm2m-iotagent/pull/248 is only failing on  the preparation step of a single test and not within the code itself. Maintaining the test suite is on TEF.

-  Ultralight - https://github.com/telefonicaid/iotagent-ul/pull/487 is open for review 
- Sigfox has no NGSI-v1 tests

**Comment by fgalan** [comment]
PR https://github.com/telefonicaid/iotagent-node-lib/pull/1054 adds some files related with NGSIv1:

* test/unit/examples/contextRequests/updateContextExpressionPlugin9.json
* test/unit/examples/contextRequests/updateContextExpressionPlugin10.json

These ones should be removed in this PR upon conflict resolution when updated with master.

**Comment by jason-fox** [comment]
Updated and conflict removed. I'm wondering why are you adding new NGSI-v1 tests - @fgalan

**Comment by fgalan** [comment]
> Updated and conflict removed. I'm wondering why are you adding new NGSI-v1 tests - @fgalan

In fact we are not adding new NGSI-v1 tests but adjusting the existing ones through the creation of new .json files (not exactly the same ;). While we still having NGIS-v1 functionality in the code, if some NGSI-v1 test breaks it seems to be a good idea to fix it to ensure regression.

**Comment by fgalan** [comment]
> Sigfox has no NGSI-v1 tests

However, IOTA Sigfox unit tests fails when using this branch (see https://github.com/telefonicaid/sigfox-iotagent/pull/110#issuecomment-880019274). How it can be?

**Comment by jason-fox** [comment]
I'd missed a couple of `/v1` mocks in the tests - fixed - https://github.com/telefonicaid/sigfox-iotagent/pull/111

**Comment by fgalan** [comment]
> @fgalan  - seeing no movement on this, I've raised two new PRs to try and reduce the scope.
>
> - If you action #1042 then no tests will run on NGSI-v1, and this PR will reduce by 189 files - coverage has temporarily reduced  by 8% with the test removal since all the legacy NGSI-v1 code remains, but it should return the same as here once this PR is finally actioned and NGSI-v1 deleted.
>
> - If you action #1043 then no references to NGSI-v1 will remain in the docs, and this PR will reduce by 7 files - coverage should remain the same as here.

Now that PR #1043 has merged, I'd suggest:

* To upgrade PR #995 with master (so the scope of the PR would be reduced to just code)
* Close PR #1042 (because at that point it should be equivalent to PR #995 and the later has all the discussion history so it's better to keep it)

@jason-fox what do you think?

**Comment by jason-fox** [comment]
Done - cba02f1 #1042 has been closed.

File: lib/plugins/attributeAlias.js [hunk 1]
```diff
@@ -93,14 +93,11 @@ function applyAlias(mappings) {
  */
 function updateAttribute(entity, typeInformation, callback) {
     const mappings = extractAllMappings(typeInformation);
-    if (config.isCurrentNgsi()) {
```
Comment (fgalan) [inline lib/plugins/attributeAlias.js:96]: We are removing usages of `isCurrentNgsi()` in IOTA agent code (e.g. PR https://github.com/telefonicaid/iotagent-ul/pull/467 PR  https://github.com/telefonicaid/iotagent-json/pull/533) so I'd suggest to remove also the `isCurrentNgsi()`  function itself in the IOTA Lib.

**Comment by fgalan** [comment]
At this point 562b6fa changes in coverage are (https://coveralls.io/builds/41398365)

imagen

File: test/unit/plugins/translation-inPlugins_test.js [hunk 2]
```diff
@@ -1,262 +0,0 @@
...
-            iotAgentLib.addQueryMiddleware(testMiddleware);
-
-            iotAgentLib.query('light1', 'Light', '', attributes, function (error, result) {
-                should.not.exist(error);
-                result.contextResponses[0].contextElement.attributes[1].value.should.equal('23%');
-                done();
-            });
-        });
-    });
-});
```
Comment (mapedraza) [inline test/unit/plugins/translation-inPlugins_test.js:262]: This file does not have an equivalent in NGSI-v2 folder, so it is needed to translate it to v2 version.

**Review by mapedraza — CHANGES_REQUESTED** [review]

File: lib/plugins/multiEntity.js [hunk 3]
```diff
@@ -342,11 +258,7 @@ function updateAttributeNgsi2(entity, typeInformation, callback) {
 }
 
 function updateAttribute(entity, typeInformation, callback) {
-    if (config.isCurrentNgsi()) {
-        updateAttributeNgsi2(entity, typeInformation, callback);
-    } else {
-        updateAttributeNgsi1(entity, typeInformation, callback);
-    }
+    updateAttributeNgsi2(entity, typeInformation, callback);
 }
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply lib/plugins/multiEntity.js]: Fixed e98d298

File: test/unit/expressions/jexlExpression-test.js [hunk 4]
```diff
@@ -1,255 +0,0 @@
...
-    describe('When an expression aims at creating an object', function () {
-        it('it should work', function (done) {
-            expressionParser.parse('{type:"Point",coordinates: [value,other]}', scope, function (error, result) {
-                should.not.exist(error);
-                result.should.deepEqual({ type: 'Point', coordinates: [6, 3] });
-                done();
-            });
-        });
-    });
-});
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply test/unit/expressions/jexlExpression-test.js]: Fixed c0dc200

File: test/unit/expressions/expression-test.js [hunk 5]
```diff
@@ -1,197 +0,0 @@
...
-    describe('When an string function is used with an expression', function () {
-        it('should work on the expression value', function (done) {
-            expressionParser.parse('indexOf(24 * @big, "80")', scope, 'String', function (error, result) {
-                should.not.exist(error);
-                result.should.equal(1);
-                done();
-            });
-        });
-    });
-});
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply test/unit/expressions/expression-test.js]: Fixed c0dc200

File: test/unit/mongodb/mongoDBUtils.js [hunk 6]
```diff
@@ -29,7 +29,7 @@ const async = require('async');
 function cleanDb(host, name, callback) {
     const url = 'mongodb://' + host + ':27017/' + name;
 
-    MongoClient.connect(url, { useNewUrlParser: true }, function (err, db) {
+    MongoClient.connect(url, { useNewUrlParser: true, useUnifiedTopology: true }, function (err, db) {
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply test/unit/mongodb/mongoDBUtils.js]: Fixed dd4c544

File: test/unit/ngsiService/active-devices-test.js [hunk 7]
```diff
@@ -1,692 +0,0 @@
...
-        });
-
-        it('should change the value of the corresponding attribute in the context broker', function (done) {
-            iotAgentLib.update('light1', 'Light', '', values, function (error) {
-                should.not.exist(error);
-                contextBrokerMock.done();
-                done();
-            });
-        });
-    });
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply test/unit/ngsiService/active-devices-test.js:196]: Fixed b966458

File: test/unit/ngsiService/active-devices-test.js [hunk 8]
```diff
@@ -1,692 +0,0 @@
...
-                contextBrokerMock.done();
-                done();
-            });
-        });
-
-        afterEach(function (done) {
-            process.env.IOTA_CB_HOST = '';
-            done();
-        });
-    });
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply test/unit/ngsiService/active-devices-test.js:691]: Fixed b966458

**Comment by jason-fox** [comment]
@mapedraza - missing test cases have been added -  this is the current diff (coveralls seems slow today)


### PR

```text
  730 passing (34s)
  11 pending

-------------------------------------|---------|----------|---------|---------|-------------------------------------------------------------------------------------------------------------------
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                                                                                 
-------------------------------------|---------|----------|---------|---------|-------------------------------------------------------------------------------------------------------------------
All files                            |      81 |    71.19 |   82.96 |   80.91 |                                                                                                                   
 lib                                 |    77.5 |    76.11 |   59.78 |   77.86 |   
```


### Master

```text
  901 passing (41s)
  14 pending

-------------------------------------|---------|----------|---------|---------|-------------------------------------------------------------------------------------------------------------------
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                                                                                 
-------------------------------------|---------|----------|---------|---------|-------------------------------------------------------------------------------------------------------------------
All files                            |   81.86 |    72.31 |   84.75 |   81.75 |                                                                                                                   
 lib                                 |    79.4 |    76.67 |   63.44 |   79.77 |     
 ```

**Comment by mapedraza** [comment]
Hi Jason, thanks for adding all those tests, we are getting close to the previous coverage! ;) 

Reviewing the latests commits, I updated my [previous comment](https://github.com/telefonicaid/iotagent-node-lib/pull/995#issuecomment-906294270) above with the list of pending tests and also resolved the change requested on the thread. However I saw a couple of subtests that still need to be added:

- On File **test/unit/ngsiv2/lazyAndCommands/lazy-devices-test.js**
   - Describe *When a context query arrives to the IoT Agent with a payload that is not JSON*
      - It 'should return an NGSI compliant payload' check is missing 

- On File **test/unit/ngsiService/queryDeviceInformationInCb-test.js**
   - Describe *When the user requests information and there are multiple responses, one of them a failure* is missing

**Comment by jason-fox** [comment]
@mapedraza - I don't think there is an equivalent NGSI-v2 request to replicate the missing test in **queryDeviceInformationInCb-test.js** - With v2 there is no such error response. if i make the **GET** request `/v2/entities/light1/attrs?attrs=state,dimming&type=Light` and the entity has`state` but not `dimming` then the response is 200 and only `state` is returned. This differs from NGSI-v1 where the body would include error information.

All the other tests are complete and the reduction is now at 0.7% - I guess much of this is removing obscure NGSI-v1 edge case cruft I wouldn't know where to look at.

**Comment by jason-fox** [comment]
@mapedraza - I don't think there is an equivalent NGSI-v2 request to replicate the missing test in **queryDeviceInformationInCb-test.js** - With v2, if i make the **GET** request `/v2/entities/light1/attrs?attrs=state,dimming&type=Light` and the entity has`state` but not `dimming` then the response is 200 and only `state` is returned. This differs from NGSI-v1 where the body would include error information.

The missing `ATTRIBUTE_NOT_FOUND` constructor can now only be fired by covering line https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/entities-NGSI-v2.js#L241 and I'm not sure if that is a legitimate NGSI-v2 response.

All the other tests are complete and the reduction remains at 0.8% - I guess much of this is obscure NGSI-v1 edge cases I wouldn't want to look at.

**Comment by fgalan** [comment]
> The missing ATTRIBUTE_NOT_FOUND constructor can now only be fired by covering line https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/entities-NGSI-v2.js#L241 and I'm not sure if that is a legitimate NGSI-v2 response.

Maybe a warning in the code should be included about this. I mean something like this:

```
            } else if (errorField.code && errorField.code === '404') {
                // FIXME: not sure if this is a legitimate case in NGSI-v2. Should be checked and removed if no longer needed
                callback(new errors.AttributeNotFound());
            } else {
```

**Comment by fgalan** [comment]
> All the other tests are complete and the reduction remains at 0.7% - I guess much of this is obscure NGSI-v1 edge cases I wouldn't want to look at.

In the case it can help, this is the current (at b1758c3) breakdown (https://coveralls.io/builds/42573501):

imagen

**Comment by jason-fox** [comment]
I'm fairly positive that the following should work:

### entities-NGSI-v2.js

```diff
         } else if (response && body && response.statusCode === 404) {
             logger.info(
                 context,
                 'Received the following response from the CB:\n\n%s\n\n',
                 JSON.stringify(body, null, 4)
             );
 
             logger.error(context, 'Operation ' + operationName + ' error connecting to the Context Broker: %j', body);
-
-            const errorField = NGSIUtils.getErrorField(body);
-            if (response.statusCode && response.statusCode === 404 && errorField !== undefined) {
+            if ( NGSIUtils.getErrorField(body) !== undefined) {
                 callback(new errors.DeviceNotFound(entityName));
-            } else if (errorField.code && errorField.code === '404') {
-                callback(new errors.AttributeNotFound());
             } else {
                 callback(new errors.EntityGenericError(entityName, typeInformation.type, body));
             }

```

### entities-NGSI-LD.js


```diff
         } else if (response && body && response.statusCode === 404) {
             logger.info(context, 'Received the following response from the CB:\n\n%s\n\n', bodyAsString);
 
             logger.error(context, 'Operation ' + operationName + ' error connecting to the Context Broker: %j', body);
 
-            const errorField = NGSIUtils.getErrorField(body);
-            if (
-                response.statusCode &&
-                response.statusCode === 404 &&
-                errorField.details.includes(typeInformation.type)
-            ) {
+            if ( NGSIUtils.getErrorField(body) !== undefined) {
                 callback(new errors.DeviceNotFound(entityName));
-            } else if (errorField.code && errorField.code === '404') {
-                callback(new errors.AttributeNotFound());
             } else {
                 callback(new errors.EntityGenericError(entityName, typeInformation.type, body));
             }
```

- The second check for `response.statusCode && response.statusCode === 404` is redundant
- `errorField.code` is *not* defined under NGSI-v2 - hence the clause is unnecessary
- NGSI-LD changes should align with NGSI-v2 - it looks like they have diverged unnecessarily.

Making this change will not break the tests. Should I add it?

**Comment by jason-fox** [comment]
Basically saying _If the broker response is a properly formed NGSI error response, then the device is not found_. If the broker response is a really nasty error with some random stack trace or something, log as a generic error.

**Comment by fgalan** [comment]
> Making this change will not break the tests. Should I add it?

Maybe after these changes the NGSIUtils.getErrorField() function can be removed? It seems a remanent usage in lib/services/devices/registrationUtils.js stills, but maybe that part can be simplified.

**Comment by jason-fox** [comment]
`NGSIUtils.getErrorField()`  **used** to be split into v1 and v2 versions and certainly can be refactored out.

**Comment by jason-fox** [comment]
Fixed 2304aae - coverage drop now 0.5% or less.  I'm assuming the remaining drop is some NGSI-v1 stuff in the IoTManger and middlewares dealing with attributes.

**Comment by fgalan** [comment]
> Fixed [2304aae](https://github.com/telefonicaid/iotagent-node-lib/commit/2304aaed427a2511adab805ae426e0457c3148f8) - coverage drop now 0.5% or less. I'm assuming the remaining drop is some NGSI-v1 stuff in the IoTManger and middlewares dealing with attributes.

Looking to the modifications in https://github.com/telefonicaid/iotagent-node-lib/compare/b1758c3...2304aae, I see getErrorField() has been removed along with two of its usages, in generateNGSI2OperationHandler() and generateNGSILDOperationHandler().

With regards to the third one I see at master within createRegistrationHandler() (https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/devices/registrationUtils.js#L57) I see you have removed the whole createRegistrationHandler() so that's fine :)

**Comment by jason-fox** [comment]
The one in `createRegistrationHandler()` would have been for NGSI-v1 and would have called the `getErrorField()` for NGSI-v1. The other two would have been for NGSI-v2 and originally called ` getErrorFieldv2()` or something. Since v1 had been deleted ` getErrorFieldv2()` was renamed to ` getErrorField()` during the refactor. It has now been deleted.

**Comment by mapedraza** [comment]
Great, all tests seems to be ported to V2. Now it is just pending a couple of pending reviews (related with nomenclature stuff):

- https://github.com/telefonicaid/iotagent-node-lib/pull/995#discussion_r671039851
- https://github.com/telefonicaid/iotagent-node-lib/pull/995#discussion_r671055058

After that, I think it is almos ready to merge

**Comment by jason-fox** [comment]
Fixed 3be10f7

**Review by mapedraza — APPROVED** [review]
LGTM

**Comment by fgalan** [comment]
The PR is in a good shape to start conducting e2e testing with IOTAs (+ Manager) generated from the following branches:

* https://github.com/telefonicaid/iotagent-json/pull/533
* https://github.com/telefonicaid/iotagent-ul/pull/467
* https://github.com/telefonicaid/sigfox-iotagent/pull/110
* https://github.com/telefonicaid/lightweightm2m-iotagent/pull/253
* https://github.com/telefonicaid/iotagent-manager/pull/238

In the meanwhile, this PR would need a CHANGES_NEXT_RELEASE entry. Suggestion:

```
- Remove: NGSI-v1 implementation (#966)
```

**Comment by jason-fox** [comment]
Fixed 4a9fd4c

**Comment by fgalan** [comment]
Good news. The CI e2e pass with the NGSIv1-removed versions went OK last night :)

So I think only this small issue is pending to do the merge: https://github.com/telefonicaid/sigfox-iotagent/pull/111#issuecomment-911631330

**Review by fgalan — APPROVED** [review]
LGTM. There is still some drop in coverage (around -0.7%) but it's acceptable (taking into account we come from a higher drop)

This PR will now be merged, followed by all the related ones in the IOTAs repositories. Thanks @jason-fox for this *huge* amount of work and also thanks to @mapedraza for his review work!
